File: horiokart2021_sensors/src/serial_odometry_node.cpp
#include "horiokart2021_sensors/serial_odometry.hpp"

#include <ros/ros.h>
#include <std_msgs/String.h>
#include <tf/transform_broadcaster.h>
#include <nav_msgs/Odometry.h>

#include <string>
#include <fstream>
#include <time.h>
#include <vector>
#include <sstream>
#include <math.h>


using namespace std;
using namespace horiokart2021_sensors;


class SerialOdometryNode
{
    private:
        ros::NodeHandle n;
        ros::NodeHandle nh_private;

        //Publisher
        ros::Publisher serial_pub;

        SerialOdometry odometry;

        string device_name;
        string odom_frame;
        string base_frame;
        int odom_pub_rate;

        bool is_write_csv;
        string csv_name;
        ofstream ofs;

        bool inv_x, inv_y, inv_th;
        bool alwaysPublish;

        const int TIME_TO_SLEEP = 100000; // usec

        ostringstream ss;

        OdometryData currentData;
        OdometryData recentNomalData;

    public:
        SerialOdometryNode();


        void init_ros_params();
        void update_odom();
        void publish_odom();

        void write_log_csv();
        void run();
};

SerialOdometryNode::SerialOdometryNode()
:nh_private("~")
{

    init_ros_params();

    if(is_write_csv){
        if(csv_name == "")
        {
            char date[64];
            time_t t = ros::WallTime::now().sec;
            strftime(date, sizeof(date), "odom_%Y%m%d%a_%H%M%S.csv", localtime(&t));
            csv_name = string(date);
            cout << date << endl;

            ROS_INFO("Odom create CSV %s", csv_name.c_str());
        }
        ofs = ofstream(csv_name);
    }
    else{
        ROS_INFO("Odom No CSV");
    }

    odometry = SerialOdometry(device_name, TIME_TO_SLEEP);
    if (!odometry.isAlive())
    {
        ROS_ERROR("Serial Fail: cound not open %s", device_name.c_str());
    }
}

void SerialOdometryNode::init_ros_params()
{
    serial_pub = n.advertise<nav_msgs::Odometry>("odom", 1);

    nh_private.param<string>("device_name", device_name, "/dev/ttyACM0");
    nh_private.param<string>("odom_frame", odom_frame, "odom");
    nh_private.param<string>("base_frame", base_frame, "base_footprint");
    nh_private.param<int>("odom_pub_rate", odom_pub_rate, 20);

    nh_private.param<bool>("inv_x", inv_x, false);
    nh_private.param<bool>("inv_y", inv_y, false);
    nh_private.param<bool>("inv_th", inv_th, false);

    nh_private.param<bool>("always_publish", alwaysPublish, true);

    nh_private.param<bool>("csv", is_write_csv, true);
    nh_private.param<string>("csv_name", csv_name, "");

}

void SerialOdometryNode::update_odom()
{
    currentData = odometry.getData();

    ss.str("");
    ss << "recv: ";
    for( const auto& b: currentData.raw)
    {
        ss << "0x" << std::setw(2) << std::setfill('0') << std::hex << static_cast<int>(b) << " ";

    }
    ROS_DEBUG("%s", ss.str().c_str());
    if (currentData.error != SerialError::NoError)
    {
        // todo switch
        ROS_ERROR("ERROR code %d", static_cast<int>(currentData.error));
    }
    else{
        if(inv_x) currentData.x *= -1;
        if(inv_y) currentData.y *= -1;
        if(inv_th) currentData.th *= -1;

        recentNomalData = currentData;
    }
}

void SerialOdometryNode::publish_odom()
{
    if(currentData.error != SerialError::NoError && !alwaysPublish)
        return;

    // double th = recentNomalData.th * 180 / 3.14;
    // double vth = recentNomalData.vth * 180 / 3.14;
    double th = recentNomalData.th ;
    double vth = recentNomalData.vth ;
    ROS_DEBUG("x:%lf y:%lf th:%lf ", recentNomalData.x, recentNomalData.y, th);
    ROS_DEBUG("vx:%lf vy:%lf vth%lf ", recentNomalData.vx, recentNomalData.vy, vth);

    geometry_msgs::Quaternion odom_quat = tf::createQuaternionMsgFromYaw(recentNomalData.th);

    ros::Time current_time = ros::Time::now();
    
    nav_msgs::Odometry odom;
    odom.header.stamp = current_time;
    odom.header.frame_id = odom_frame;

    odom.pose.pose.position.x = recentNomalData.x;
    odom.pose.pose.position.y = recentNomalData.y;
    odom.pose.pose.position.z = 0.0;
    odom.pose.pose.orientation = odom_quat;

    double vx = sqrt(pow(recentNomalData.vx, 2) + pow(recentNomalData.vy, 2));
    odom.child_frame_id = base_frame;
    // odom.twist.twist.linear.x = recentNomalData.vx;
    // odom.twist.twist.linear.y = recentNomalData.vy;
    odom.twist.twist.linear.x = vx;
    odom.twist.twist.linear.y = 0;
    odom.twist.twist.angular.z = vth;

    serial_pub.publish(odom);
}

void SerialOdometryNode::write_log_csv()
{
    if(is_write_csv){

        ofs << ros::WallTime::now().sec << "." << ros::WallTime::now().nsec << ",";
        ofs << currentData.x << ",";
        ofs << currentData.y << ",";
        ofs << currentData.th << ",";
        ofs << currentData.vx << ",";
        ofs << currentData.vy << ",";
        ofs << currentData.vth << ",";
        ofs << ",";
        ofs << static_cast<int>(currentData.error) << ",";
        ofs << ",";
        for(auto& b:currentData.raw){
            ofs << static_cast<int>(b) << ",";
        }
        ofs << endl;
    }

}

void SerialOdometryNode::run()
{
    ROS_INFO("start serial odom");

    SerialError e = odometry.sendZeroReset(3);
    if (e == SerialError::NoError)
    {
        ROS_INFO("Odom zero reset");
    }
    else{
        ROS_ERROR("Odom error zero reset!");
    }

    ros::Rate loop_rate(odom_pub_rate); 
    while (ros::ok()){

        update_odom();
        publish_odom();
        write_log_csv();

        ros::spinOnce();
        loop_rate.sleep();
    } 
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "serial_odometry");

    SerialOdometryNode s_odom;
    s_odom.run();
}

File: horiokart2021_sensors/script/serial_odometry_node.py
#!/usr/bin/env python3
import rospy
import tf
from tf.transformations import quaternion_from_euler
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Quaternion
import math
import datetime
import csv
import os
import time

# from horiokart2021_sensors.serial_odometry import SerialOdometry
from horiokart2021_sensors.src.horiokart2021_sensors.serial_odometry import SerialOdometry, SerialCommand, OdometryData
from horiokart2021_sensors.src.horiokart2021_sensors.serial_communicator import SerialCommand, SerialCommunicator, SerialError


class SerialOdometryNode():
    def __init__(self) -> None:

        self._odom_frame = rospy.get_param("~odom_frame", "odom")
        self._base_frame = rospy.get_param("~base_frame", "base_footprint")
        self._pub_rate = rospy.get_param("~odom_pub_rate", 20)

        self._inv_x = rospy.get_param("~inv_x", False)
        self._inv_y = rospy.get_param("~inv_y", False)
        self._inv_th = rospy.get_param("~inv_th", False)

        self._always_publish = rospy.get_param("~always_publis", True)

        self._is_write_csv = rospy.get_param("~csv", True)
        self._csv_path = rospy.get_param("~csv_path", "/root/ros_data/log_csv")
        self._csv_name = rospy.get_param("~csv_name", "")

        if self._is_write_csv:
            if self._csv_name == "":
                dt_now = datetime.datetime.now()
                self._csv_name = dt_now.strftime("odom_%Y%m%d%a_%H%M%S.csv")

            with open(os.path.join(self._csv_path,self._csv_name), "w") as _:
                rospy.loginfo(f"Odom create CSV {os.path.join(self._csv_path,self._csv_name)}")
            
        else:
            rospy.loginfo(f"Odom No CSV")


        device_name = rospy.get_param("~device_name", "ttyHoriokart-odom")
        baud = rospy.get_param("~baud", 56700)
        timeout = rospy.get_param("~timeout", 0.5)

        self._serial = SerialCommunicator(
            device_name=device_name,
            baud=baud,
            timeout=timeout
        )
        self._odom = SerialOdometry(communicator=self._serial)

        self._odom_pub = rospy.Publisher(
            "odom",
            Odometry,
            queue_size=1
        )

        self._current_data: OdometryData = None
        self._recent_valid_data: OdometryData = None


    def update_odom(self):
        self._current_data = self._odom.get_odom()

        rospy.logdebug(f"read:{''.join([f'0x{x:02x} ' for x in self._current_data.raw])}")

        if self._current_data.error != SerialError.NoError:
            rospy.logerr(f"ERROR code {self._current_data.error}")

        else:
            if self._inv_x:
                self._current_data.x *= -1
            if self._inv_y:
                self._current_data.y *= -1
            if self._inv_th:
                self._current_data.th *= -1

            self._recent_valid_data = self._current_data

    def publish(self):
        if (
            self._current_data.error != SerialError.NoError
                and not self._always_publish
                or self._recent_valid_data is None
        ):
            return

        rospy.logdebug(f"x: {self._recent_valid_data.x:.3f} y: {self._recent_valid_data.y:.3f} th: {self._recent_valid_data.th:.3f}")
        rospy.logdebug(f"vx: {self._recent_valid_data.vx:.3f} vy: {self._recent_valid_data.vy:.3f} vth: {self._recent_valid_data.vth:.3f}")

        odom_msg = Odometry()

        odom_msg.header.stamp = rospy.Time().now()
        odom_msg.header.frame_id = self._odom_frame
        odom_msg.child_frame_id = self._base_frame

        odom_msg.pose.pose.position.x = self._recent_valid_data.x
        odom_msg.pose.pose.position.y = self._recent_valid_data.y
        odom_msg.pose.pose.position.z = self._recent_valid_data.z

        odom_quat = quaternion_from_euler(0, 0, self._recent_valid_data.th)
        odom_msg.pose.pose.orientation = odom_quat

        odom_msg.twist.twist.linear.x = math.sqrt(
            self._recent_valid_data.vx**2 + self._recent_valid_data.vy**2
        )
        odom_msg.twist.twist.linear.y = 0
        odom_msg.twist.twist.angular.z = self._recent_valid_data.vth

        self._odom_pub.publish(odom_msg)

    def write_log_csv(self):
        if self._is_write_csv:
            with open(os.path.join(self._csv_path, self._csv_name)) as f:
                writer = csv.writer(f)

                writer.writerow(
                    [
                        time.time(), # epoch milli sec
                        self._current_data.x,
                        self._current_data.y,
                        self._current_data.th,
                        self._current_data.vx,
                        self._current_data.vy,
                        self._current_data.vth,
                        "",
                        self._current_data.error,
                        "",
                        *[f'0x{x:02x}' for x in self._current_data.raw]
                    ]
                )


    def run(self):
        rospy.loginfo("start serial odom node")

        e = self._odom.zero_reset(retry=3)
        if e == SerialError.NoError:
            rospy.loginfo("Odometry zero reset DONE")
        else:
            rospy.logwarn("Odometry zero reset ERROR")

        rate = rospy.Rate(self._pub_rate)
        while not rospy.is_shutdown():
            self.update_odom()

            self.publish()

            # todo: write csv

            rate.sleep()


if __name__ == "__main__":
    rospy.init_node("serial_odometry_node")

    node = SerialOdometryNode()
    node.run()
